### PR TITLE
Test case verbosity

### PR DESCRIPTION
**Pull request by plannigan**
Allow for the output of test case execution to be controlled independently from the application verbosity level. `verbosity_test_case` is the new ini setting to adjust this functionality.

Addresses #11639

### CODE REVIEW

File: changelog/11639.feature.rst
Comment (nicoddemus) [inline changelog/11639.feature.rst:1]: The "failed assertions verbosity" part at the end seems copied and paste from the another entry.

In fact, given that this was not released yet, I think it makes more sense to just update the other entry to mention both options together.

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:301]:
```suggestion
:confval:`verbosity_test_cases`: Controls how verbose the test execution output should be when pytest is executed.
```

File: changelog/11639.feature.rst
Comment (nicoddemus) [inline changelog/11639.feature.rst:1]:
```suggestion
Added the new :confval:`verbosity_test_cases` configuration option for fine-grained control of failed assertions verbosity.
```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1838]:
```suggestion
.. confval:: verbosity_test_cases
```

File: doc/en/reference/reference.rst
Comment (nicoddemus) [inline doc/en/reference/reference.rst:1845]:
```suggestion
        verbosity_test_cases = 2
```

File: testing/test_terminal.py
Comment (nicoddemus) [inline testing/test_terminal.py:2619]: Up to you, but when I have "groups" of tests like this I prefer to group them explicitly using a class, so beside being explicit we can use shorter names for the tests:

```python
def test_fine_grained_test_case_verbosity(pytester: Pytester): ...
def test_fine_grained_test_case_verbosity_collect_only_negative_2(pytester: Pytester): ...
def test_fine_grained_test_case_verbosity_collect_only_positive_2(pytester: Pytester): ...
```

vs:

```python
class TestFineGrainedTestCase:
    def test_verbosity(self, pytester: Pytester): ...
    def test_collect_only_negative_2(self, pytester: Pytester): ...
    def test_collect_only_positive_2(self, pytester: Pytester): ...
```

As a bonus the `_fine_grained_verbosity_file_contents` utility can be turned into a method of the class, making it more context, and can be renamed too to something simpler like `file_contents()` (no need for the `_`).

Comment (nicoddemus) [inline testing/test_terminal.py:2619]:
```suggestion
def test_fine_grained_test_case_verbosity(pytester: Pytester) -> None:
```

File: testing/test_terminal.py
Comment (nicoddemus) [inline testing/test_terminal.py:2645]:
```suggestion
def test_fine_grained_test_case_verbosity_collect_only_negative_2(pytester: Pytester) -> None:
```

File: testing/test_terminal.py
Comment (nicoddemus) [inline testing/test_terminal.py:2665]:
```suggestion
def test_fine_grained_test_case_verbosity_collect_only_positive_2(pytester: Pytester) -> None:
```

File: testing/test_terminal.py
Comment (nicoddemus) [inline testing/test_terminal.py:2695]: Not sure we need this complex tests, just a simple parametrized test would allow us to test the purpose of the flag just as well, right?

```python
@pytest.mark.parametrize("i", range(7))
def test_ok(i): pass
```

The rationale is that we are not testing the details of the output, only that it is more or less verbose, so the names and the length of the tests docstrings do not plan a role here, and they are already being tested elsewhere.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks @plannigan for following up with this. 👍 

Please take a look at my comments.

One thing I'm not entirely happy is with the "test cases" name, as that IMHO does not convey at all what the flag does. On the other hand I don't have a good suggestion to provide. Would really like some input from other maintainers here (@bluetech @RonnyPfannschmidt @The-Compiler).

**Review by nicoddemus — APPROVED** [review]
LGTM!

Will leave it open for a few days to give others a chance to chime in -- perhaps someone will come up with a great suggestion for the option name. 🙏

**Comment by bluetech** [comment]
@plannigan In https://github.com/pytest-dev/pytest/issues/11387#issuecomment-1722324799 you did some great work listing all of the places where verbosity is used. While the assertions verbosity level was a common feature request, I'm not *entirely* sure about the others (as opposed to using the global verbosity level).

For me, in order to properly evaluate going forward, it will be helpful if you can list all of the categories you're thinking about and how the uses listed in #11387 will categorize.

**Comment by bluetech** [comment]
> For "test case" specifically, I've seen people wanting to use more verbosity for it, while keeping using less verbosity for assertions (it was internally at work on a specific case), so I think this option has its uses.

Right, but for this they can bump `verbosity` and decrease `verbosity_assertions` now. What I'm curious about is whether there's demand for controlling the remaining cases besides assertions separately.

**Comment by plannigan** [comment]
@bluetech I think it is a good idea to take a step back and consider where this work could go. After reviewing my original list of where `verbosity` is used and things seem to fall into 3 larger categories:

* Things in the header
* Things in the summary (this covers assertions, which has already been merged)
* The things in between (this PR)

There is one other potential category dealing with traceback filtering and truncating locals. Since that is in the summary, there might be an argument that it should also be grouped with assertions, but at first glance, it feels distinct to me. (Maybe if `verbosity_assertions` was renamed to `verbosity_summary` I might think differently.)

As for the "would people use this?" question. I agree that people could achieve it as described. But I feel that having this being the answer is:
* harder for people to discover
* a bit clunky (`addopts = "-vv"` and `verbosity_assertions = 0` to do everything in the config file).
* less explicit - a reader would have to understand the implications of both being used. Going further, the usages might not be next to each other, making it harder to know both are used.

**Comment by plannigan** [comment]
@bluetech @nicoddemus Checking back in as there hasn't been any activity for 2 weeks since my last reply (hopefully you both took some time off for the holidays).

**Comment by nicoddemus** [comment]
> (hopefully you both took some time off for the holidays).

Indeed this was the case for me. 

But the remaining discussion is probably best continued by @bluetech

**Comment by plannigan** [comment]
@bluetech I have updated the branch so that it would be able to be included in the 8.0 release with `verbosity_assertions`. Do you have any more requests/feedback or can this move forward?

**Comment by bluetech** [comment]
@plannigan Sorry for not responding earlier and letting it linger -- not very nice.

I read [your comment](https://github.com/pytest-dev/pytest/pull/11653#issuecomment-1866853616) but I am not convinced personally. I think it is better to only add configuration knobs when there is demand for it, and my feeling is that there isn't clear demand for this one. I don't mean to say that the feature isn't useful per se, just that IMO it's not useful *enough*, it's a balancing act.

So I'm -0 on this feature, meaning I won't merge it myself but if any other maintainer feels otherwise that would be perfectly fine by me.

**Comment by nicoddemus** [comment]
If @bluetech is -0 on this, I would like to get this merged then. This is something people at work have mentioned in the past so I think it is a nice addition.

I will make a final review tomorrow and merge it then. 👍

**Comment by plannigan** [comment]
Thanks for the responses. I've updated the branch to include the latest of main and reverted that changelog changes I had made to be compatible with the v8 rc. Should be good to go.

**Review by nicoddemus — APPROVED** [review]
Again thanks @plannigan for all the work on this. 👍

**Pull request merged**
